runtime/src/main/java/org/corfudb/runtime/collections/streaming/StreamingTask.java

```diff
@@ -0,0 +1,188 @@
+                                        "have specified stream tag: %s", t.getFullyQualifiedTableName(), streamTag));
+                            }
+                            return new TableSchema<>(tName, t.getKeyClass(), t.getValueClass(), t.getMetadataClass());
+                        }));
+        this.status = new AtomicReference<>(StreamStatus.RUNNABLE);
+    }
+
+    public StreamStatus getStatus() {
+        return status.get();
+    }
+
+    public void move(StreamStatus from, StreamStatus to) {
+        Preconditions.checkState(status.compareAndSet(from, to),
+                "move: failed to change %s to %s", from, to);
+    }
+
+    private Optional<CorfuStreamEntries> transform(ILogData logData) {
+        Objects.requireNonNull(logData);
+        // Avoid LogData de-compression if it does not contain any table of interest.
+        if (logData.isHole() || Sets.intersection(logData.getStreams(), tableSchemas.keySet()).isEmpty()) {
+            return Optional.empty();
+        }
+
+        long epoch = logData.getEpoch();
+        MultiObjectSMREntry smrEntries = (MultiObjectSMREntry) logData.getPayload(runtime);
+
+        Map<UUID, TableSchema<K, V, M>> filteredSchemas = logData.getStreams()
+                .stream()
+                .filter(tableSchemas::containsKey)
+                .collect(Collectors.toMap(Function.identity(), tableSchemas::get));
+
+        Map<TableSchema, List<CorfuStreamEntry>> streamEntries = new HashMap<>();
+
+        filteredSchemas.forEach((streamId, schema) -> {
+            // Only deserialize the interested streams to reduce overheads.
+            List<CorfuStreamEntry> entryList = smrEntries.getSMRUpdates(streamId)
+                    .stream()
+                    .map(entry -> CorfuStreamEntry.fromSMREntry(entry, epoch))
+                    .collect(Collectors.toList());
+
+            // Deduplicate entries per stream Id, ordering within a transaction is not guaranteed
+            Map<Message, CorfuStreamEntry> observedKeys = new HashMap<>();
+            entryList.forEach(entry -> observedKeys.put(entry.getKey(), entry));
+
+            if (!entryList.isEmpty()) {
+                streamEntries.put(schema, new ArrayList<>(observedKeys.values()));
+            }
+        });
+
+        // This transaction data does not contain any table of interest, don't add to buffer.
+        if (streamEntries.isEmpty()) {
+            return Optional.empty();
+        }
+
+        CorfuStoreMetadata.Timestamp timestamp = CorfuStoreMetadata.Timestamp.newBuilder()
+                .setSequence(logData.getGlobalAddress())
+                .setEpoch(epoch)
+                .build();
+
+        return Optional.of(new CorfuStreamEntries(streamEntries, timestamp));
+    }
+
+    public DeltaStream getStream() {
+        return this.stream;
+    }
+
+    private void produce() {
+        Preconditions.checkState(status.get() == StreamStatus.SYNCING);
+        Preconditions.checkState(stream.hasNext());
```

Comment:
Yes, this is by design because the scheduler will not dispatch this task to the worker pool unless `hasNext` returns true. If this precondition fails it means there is a concurrency issue and I would like it to explicitly fail. 

```java

private void dispatchSyncTasks(List<StreamingTask> tasks) {
        for (StreamingTask task : tasks) {
            if (task.getStream().hasNext()) {
                // this will fail for tasks that are already syncing
                // for tasks to be scheduled to sync, they must be in the scheduling state
                task.move(StreamStatus.SCHEDULING, StreamStatus.SYNCING);
                workers.execute(task);
            } else {
                // no new deltas to sync
                // This stream has no entries to sync after the poll, set it to runnable so that it can
                // be polled next scheduling cycle
                task.move(StreamStatus.SCHEDULING, StreamStatus.RUNNABLE);
            }
        }
    }

```